pkgs/os-specific/linux/systemd/0020-timesyncd-disable-NSCD-when-DNSSEC-validation-is-dis.patch

```diff
@@ -0,0 +1,59 @@
+From fc5a9163b7489cf601aca1bf20a9b51340b7362f Mon Sep 17 00:00:00 2001
+From: Yuri Nesterov <yuriy.nesterov@unikie.com>
+Date: Wed, 21 Jun 2023 17:17:38 +0300
+Subject: [PATCH] timesyncd: disable NSCD when DNSSEC validation is disabled
+
+Systemd-timesyncd sets SYSTEMD_NSS_RESOLVE_VALIDATE=0 in the unit file
+to disable DNSSEC validation but it doesn't work when NSCD is used in
+the system. This patch disabes NSCD in systemd-timesyncd when
+SYSTEMD_NSS_RESOLVE_VALIDATE is set to 0 so that it uses NSS libraries
+directly.
+---
+ src/timesync/timesyncd.c           | 11 +++++++++++
+ units/systemd-timesyncd.service.in |  1 +
+ 2 files changed, 12 insertions(+)
+
+diff --git a/src/timesync/timesyncd.c b/src/timesync/timesyncd.c
+index 1d8ebecc91..2b0ae361ff 100644
+--- a/src/timesync/timesyncd.c
++++ b/src/timesync/timesyncd.c
+@@ -21,6 +21,11 @@
+ #include "timesyncd-conf.h"
+ #include "timesyncd-manager.h"
+ #include "user-util.h"
++#include "env-util.h"
++
++struct traced_file;
++extern void __nss_disable_nscd(void (*)(size_t, struct traced_file *));
++static void register_traced_file(size_t dbidx, struct traced_file *finfo) {}
+ 
+ static int advance_tstamp(int fd, const struct stat *st) {
+         assert_se(fd >= 0);
+@@ -198,6 +203,12 @@ static int run(int argc, char *argv[]) {
+         if (r < 0)
+                 return log_error_errno(r, "Failed to parse fallback server strings: %m");
+ 
++        r = getenv_bool_secure("SYSTEMD_NSS_RESOLVE_VALIDATE");
++        if (r == 0) {
++                log_info("Disabling NSCD because DNSSEC validation is turned off");
++                __nss_disable_nscd(register_traced_file);
++        }
++
+         log_debug("systemd-timesyncd running as pid " PID_FMT, getpid_cached());
+ 
+         notify_message = notify_start("READY=1\n"
+diff --git a/units/systemd-timesyncd.service.in b/units/systemd-timesyncd.service.in
+index c606461091..d5789188ee 100644
+--- a/units/systemd-timesyncd.service.in
++++ b/units/systemd-timesyncd.service.in
+@@ -26,6 +26,7 @@ CapabilityBoundingSet=CAP_SYS_TIME
+ # correct time to work, but we likely won't acquire that without NTP. Let's
+ # break this chicken-and-egg cycle here.
+ Environment=SYSTEMD_NSS_RESOLVE_VALIDATE=0
++Environment=LD_LIBRARY_PATH={{ROOTLIBDIR}}
```

Comment:
We only need to start timesyncd with `LD_LIBRARY_PATH` set like this if systemd-resolved itself is also enabled (`services.resolved.enable = true;`) too, right?

Otherwise nothing will honor the `SYSTEMD_NSS_RESOLVE_VALIDATE` env var that's set.

I think I'd prefer if we set this in the NixOS module (in `nixos/modules/system/boot/timesyncd.nix`), and only if resolved is enabled, rather than patching it in uncondtionally:

```
systemd.services.systemd-timesyncd.environment.LD_LIBRARY_PATH = if config.services.resolved.enable "${config.systemd.package.out}/lib" else "";
```

The other part of the patch should obviously stay as-is.

Reply:
Actually, it looks like with the nssModules path added to systemd-timesyncd, everything works fine even without the patch because when NSCD fails to resolve the hostname, it falls back to modules from nsswitch.conf. In this case, systemd-resolved resolves the hostname fine because when loaded directly from timesyncd, it gets SYSTEMD_NSS_RESOLVE_VALIDATE=0. Disabling NSCD in systemd-timesyncd helps to save a few seconds, but now I'm not sure if it's worth it.

Reply:
The current approach is what I suggested, no? Keep the SYSTEMD_NSS_RESOLVE_VALIDATE=0 part in the patch, but configure the LD_LIBRARY_PATH env var through the module system.

Reply:
Yes, but my point is that if we use "if config.services.resolved.enable" then when resolved is not enabled, systemd-timesyncd will not be able to use any other NSS modules because NSCD is disabled in the patch and NSS modules are not in LD_LIBRARY_PATH.

Reply:
But nss-resolve is the only NSS module honoring `SYSTEMD_NSS_RESOLVE_VALIDATE=0` env var anyways, no?

Reply:
That's correct. I'm just saying that when nss-resolve is disabled we probably still want systemd-timesyncd to keep using other NSS modules that are configured in the system.

Reply:
Oh yes indeed! So `config.system.nssModules.path` makes more sense.

Can you extend the comment in the module system to make that point more clear?

Reply:
Sure, I tried to explain in the comment.

Reply:
LGTM!